src/hooks/useVisibleSection.ts

```diff
@@ -0,0 +1,49 @@
+//Credits: https://medium.com/@guilherme.ramalho/creating-a-side-menu-with-section-highlighting-in-react-8247be6c9f56
+'use client';
+
+import { useCallback, useEffect, useState } from 'react';
+import { useMounted } from '@/hooks/useMounted';
+import { useThrottle } from '@/hooks/useThrottle';
+
+export function useVisibleSection(sections: string[]) {
+  const isMounted = useMounted();
+  const [visibleSectionId, setVisibleSectionId] = useState<
+    string | undefined
+  >();
+
+  const isSectionVisible = (elementId: string) => {
+    const section = document.getElementById(elementId);
+
+    if (section) {
+      const sectionPosition = section.getBoundingClientRect();
+
+      return (
+        sectionPosition.top >= 0 && sectionPosition.bottom <= window.innerWidth
+      );
+    }
+
+    return false;
+  };
+
+  const checkVisibility = useThrottle(
+    useCallback(() => {
+      sections.forEach((id) => {
+        if (isSectionVisible(id)) {
+          setVisibleSectionId(id);
+        }
+      });
```

Comment:
Use modern iterator:
```suggestion
      for (const id of sections)
        if (isSectionVisible(id)) setVisibleSectionId(id);
```